### ISSUE TITLE
Inquiry VideoLLaMA2-7B Model Performance Discrepancies

### ISSUE THREAD

**Issue by chuyq** [DAMO-NLP-SG/VideoLLaMA2#34]
Thanks for releasing the code for the great work.
I utilized the VideoLLaMA2-7B (chat) model for offline reasoning and noted significant discrepancies compared to its online performance using the same prompt.

The prompt instructed: "Carefully observe the facial expressions of the individuals involved. Select the most appropriate emotional state from the following options: anger, disgust, sadness, joy, neutral. Choose one or two emotions." 

The offline output is: "The video shows a man sitting in a chair talking to a woman. Both are dressed in black, seated in a room with wooden and white doors, alongside brown leather furniture. The video lasts 8 seconds." 
Conversely, the online response is "sadness, depression, fear."

Could this difference be due to different configuration parameters used during offline and online inference? If so, what parameters would adjusting in the inference file align the offline results with the online demo's performance?

**Comment by sennnnn** [comment]
We have fix this Discrepancies. Please check this [commit](https://github.com/DAMO-NLP-SG/VideoLLaMA2/commit/c53d1d84cc89157e1af2df45a6c3ab19a480da8a).

**Comment by sennnnn** [comment]
Ok, this is a modified inference script which matches the pipeline of gradio demo:
```python
import torch
import transformers

import sys
sys.path.append('./')
from videollama2.conversation import conv_templates
from videollama2.constants import DEFAULT_MMODAL_TOKEN, MMODAL_TOKEN_INDEX
from videollama2.mm_utils import get_model_name_from_path, tokenizer_MMODAL_token, process_video, process_image
from videollama2.model.builder import load_pretrained_model


def inference():
    # Video Inference
    paths = ['assets/cat_and_chicken.mp4']
    questions = ['What animals are in the video, what are they doing, and how does the video feel?']
    # Reply:
    # The video features a kitten and a baby chick playing together. The kitten is seen laying on the floor while the baby chick hops around. The two animals interact playfully with each other, and the video has a cute and heartwarming feel to it.
    modal_list = ['video']

    # 1. Initialize the model.
    model_path = 'DAMO-NLP-SG/VideoLLaMA2-7B-16F'
    model_name = get_model_name_from_path(model_path)
    tokenizer, model, processor, context_len = load_pretrained_model(model_path, None, model_name)
    model = model.to('cuda:0')
    conv_mode = 'llama_2'

    # 2. Visual preprocess (load & transform image or video).
    if modal_list[0] == 'video':
        tensor = process_video(paths[0], processor, model.config.image_aspect_ratio, num_frames=16, sample_scheme='fps').to(dtype=torch.float16, device='cuda', non_blocking=True)
        default_mm_token = DEFAULT_MMODAL_TOKEN["VIDEO"]
        modal_token_index = MMODAL_TOKEN_INDEX["VIDEO"]
    else:
        tensor = process_image(paths[0], processor, model.config.image_aspect_ratio)[0].to(dtype=torch.float16, device='cuda', non_blocking=True)
        default_mm_token = DEFAULT_MMODAL_TOKEN["IMAGE"]
        modal_token_index = MMODAL_TOKEN_INDEX["IMAGE"]
    tensor = [tensor]

    # 3. text preprocess (tag process & generate prompt).
    question = default_mm_token + "\n" + questions[0]
    conv = conv_templates[conv_mode].copy()
    conv.append_message(conv.roles[0], question)
    conv.append_message(conv.roles[1], None)
    prompt = conv.get_prompt()
    input_ids = tokenizer_MMODAL_token(prompt, tokenizer, modal_token_index, return_tensors='pt').unsqueeze(0).to('cuda:0')

    with torch.inference_mode():
        output_ids = model.generate(
            input_ids,
            images_or_videos=tensor,
            modal_list=modal_list,
            do_sample=True,
            temperature=0.2,
            max_new_tokens=1024,
            use_cache=True,
        )

    outputs = tokenizer.batch_decode(output_ids, skip_special_tokens=True)
    print(outputs[0])


if __name__ == "__main__":
    inference()
```

Hope this script can help you.

**Comment by chuyq** [comment]
This is no different from the previous code.

**Comment by sennnnn** [comment]
> This is no different from the previous code.

Do your modify the argument of process_video? The `sample_scheme='fps'` and `num_frames=16`.

**Comment by chuyq** [comment]
Yes, I make the changes, but it still doesn't work; thank you. The results remain as follows:
![image](https://github.com/DAMO-NLP-SG/VideoLLaMA2/assets/87362339/6c7c8b43-caa2-4ea4-bb35-da13def989c3)

**Comment by sennnnn** [comment]
Is is possible to provide some cases for testing our inference code? We need to reproduce your bug.

**Comment by chuyq** [comment]
Certainly! Thank you,  I'll send you sample test videos along with the inference code in the email.

**Comment by sennnnn** [comment]
We have tested your code and find that the input prompt has format bug. The `questions = question` cause incomplete input prompt. You should set `questions = [question]` in your code.

**Comment by LiangMeng89** [comment]
> Thanks for releasing the code for the great work. I utilized the VideoLLaMA2-7B (chat) model for offline reasoning and noted significant discrepancies compared to its online performance using the same prompt.
> 
> The prompt instructed: "Carefully observe the facial expressions of the individuals involved. Select the most appropriate emotional state from the following options: anger, disgust, sadness, joy, neutral. Choose one or two emotions."
> 
> The offline output is: "The video shows a man sitting in a chair talking to a woman. Both are dressed in black, seated in a room with wooden and white doors, alongside brown leather furniture. The video lasts 8 seconds." Conversely, the online response is "sadness, depression, fear."
> 
> Could this difference be due to different configuration parameters used during offline and online inference? If so, what parameters would adjusting in the inference file align the offline results with the online demo's performance?

Hello,I'm a phD student from ZJU, I also use videollama2 to do my own research,we create a WeChat group to discuss some issues of videollama2 and help each other,could you join us? Please contact me: WeChat number == LiangMeng19357260600, phone number == +86 19357260600,e-mail == liangmeng89@zju.edu.cn.
